File: src/analytics.ts
/** Aggregations over the decision log for `tiershift report` and `tiershift tune`. Pure functions. */
import { applyPolicy, estimateOutputTokens } from "./policy.js";
import type { LogEntry } from "./log.js";
import type { Config, ModelMeta } from "./types.js";

export interface TierStat { tier: string; n: number; share: number; cost: number; est_cost: number; mean_confidence: number; p50_jev_ms: number }
export interface Report {
  n: number; from: string | null; to: string | null;
  tiers: TierStat[];
  models: { model: string; n: number; cost: number; est_cost: number }[];
  total_cost: number; total_est_cost: number;
  /** What the same requests would have cost on the top model of the top tier, by estimated tokens. */
  flagship_est_cost: number; flagship_model: string | null;
  saving_vs_flagship: number | null;
  jev: { p50_ms: number; p95_ms: number; cost: number };
  degraded: number; fell_back: number; low_confidence: number;
  overrides: { reason: string; n: number }[];
}

const pct = (xs: number[], p: number) => { const s = [...xs].sort((a, b) => a - b); return s.length ? s[Math.min(s.length - 1, Math.floor(p * s.length))] : 0; };
const sum = (xs: number[]) => xs.reduce((a, b) => a + b, 0);
const JEV_PRICE = 0.042 / 1e6;

export function estCostFor(meta: ModelMeta | undefined, inTok: number, outTok: number): number | null {
  if (!meta?.price) return null;
  return (inTok * meta.price.input + outTok * meta.price.output) / 1e6;
}

/** Cost of one entry: actual when a model was called, else the estimate. */
export function entryCost(e: LogEntry): number { return e.cost_usd ?? e.est_cost_usd ?? 0; }

export interface AnalyticsOptions {
  /** Tier → models actually usable (see `availableTiers`). Defaults to the configured lists. */
  tiers?: Record<string, string[]>;
  /** Model for the always-flagship comparison. Defaults to the first usable model of the top tier. */
  baselineModel?: string;
}

/** First usable model of a tier. */
export function modelForTier(config: Config, tier: string, tiers?: Record<string, string[]>): string | undefined {
  return (tiers ?? config.tiers)[tier]?.[0];
}

export function buildReport(entries: LogEntry[], config: Config, opts: AnalyticsOptions = {}): Report {
  const order = Object.keys(config.tiers);
  const flagshipTier = order[order.length - 1];
  const flagshipModel = opts.baselineModel ?? modelForTier(config, flagshipTier, opts.tiers) ?? null;
  const tiers: TierStat[] = order.map((tier) => {
    const es = entries.filter((e) => e.tier === tier);
    return { tier, n: es.length, share: es.length / (entries.length || 1), cost: sum(es.map(entryCost)), est_cost: sum(es.map((e) => e.est_cost_usd ?? 0)), mean_confidence: es.length ? sum(es.map((e) => e.confidence)) / es.length : 0, p50_jev_ms: pct(es.map((e) => e.jev_latency_ms), 0.5) };
  });
  const modelIds = [...new Set(entries.map((e) => e.model))].sort();
  const models = modelIds.map((model) => { const es = entries.filter((e) => e.model === model); return { model, n: es.length, cost: sum(es.map(entryCost)), est_cost: sum(es.map((e) => e.est_cost_usd ?? 0)) }; });
  const flagshipEst = flagshipModel ? sum(entries.map((e) => estCostFor(config.models?.[flagshipModel], e.code_signals.est_input_tokens, e.output_tokens ?? estimateOutputTokens(e.signals.output_length)) ?? 0)) : 0;
  const total = sum(entries.map(entryCost));
  const overrideCounts = new Map<string, number>();
  for (const e of entries) for (const r of e.reason) if (r.startsWith("override")) overrideCounts.set(r.split(" → ")[0], (overrideCounts.get(r.split(" → ")[0]) ?? 0) + 1);
  return {
    n: entries.length, from: entries[0]?.ts ?? null, to: entries[entries.length - 1]?.ts ?? null,
    tiers, models, total_cost: total, total_est_cost: sum(entries.map((e) => e.est_cost_usd ?? 0)),
    flagship_est_cost: flagshipEst, flagship_model: flagshipModel,
    saving_vs_flagship: flagshipEst > 0 ? 1 - total / flagshipEst : null,
    jev: { p50_ms: pct(entries.map((e) => e.jev_latency_ms), 0.5), p95_ms: pct(entries.map((e) => e.jev_latency_ms), 0.95), cost: sum(entries.map((e) => e.jev_input_tokens * JEV_PRICE)) },
    degraded: entries.filter((e) => e.degraded).length, fell_back: entries.filter((e) => e.fell_back).length,
    low_confidence: entries.filter((e) => e.confidence < 0.5).length,
    overrides: [...overrideCounts.entries()].map(([reason, n]) => ({ reason, n })).sort((a, b) => b.n - a.n),
  };
}

export interface TuneResult {
  n: number;
  baseline: { tiers: Record<string, number>; est_cost: number };
  candidate: { tiers: Record<string, number>; est_cost: number };
  moved_down: number; moved_up: number; unchanged: number;
  saving: number | null;
  /** Entries where the candidate's rules alone would pick a different tier, but an override held the result. */
  held_by_override: number;
  /** Entries that would move, with the old and new tier, for spot checks. */
  moves: { ts: string; from: string; to: string; difficulty: number; stakes: number; confidence: number }[];
}

/** Replay logged signals against a candidate config. No Jev calls: the signals are already in the log. */
export function tune(entries: LogEntry[], baseline: Config, candidate: Config, opts: AnalyticsOptions = {}): TuneResult {
  const order = Object.keys(candidate.tiers);
  const count = (tiers: string[]) => Object.fromEntries(order.map((t) => [t, tiers.filter((x) => x === t).length]));
  const cost = (cfg: Config, e: LogEntry, tier: string) => {
    const model = modelForTier(cfg, tier, opts.tiers);
    return estCostFor(model ? cfg.models?.[model] : undefined, e.code_signals.est_input_tokens, e.output_tokens ?? estimateOutputTokens(e.signals.output_length)) ?? 0;
  };
  const rulesOnly = (cfg: Config): Config => ({ ...cfg, overrides: [] });
  const base: string[] = [], cand: string[] = [], moves: TuneResult["moves"] = [];
  let bCost = 0, cCost = 0, down = 0, up = 0, held = 0;
  for (const e of entries) {
    const b = applyPolicy(baseline, e.signals, e.code_signals).tier;
    const c = applyPolicy(candidate, e.signals, e.code_signals).tier;
    base.push(b); cand.push(c); bCost += cost(baseline, e, b); cCost += cost(candidate, e, c);
    if (b !== c) { if (order.indexOf(c) < order.indexOf(b)) down++; else up++; moves.push({ ts: e.ts, from: b, to: c, difficulty: e.signals.difficulty, stakes: e.signals.stakes, confidence: e.confidence }); }
    else if (applyPolicy(rulesOnly(baseline), e.signals, e.code_signals).tier !== applyPolicy(rulesOnly(candidate), e.signals, e.code_signals).tier) held++;
  }
  return { n: entries.length, baseline: { tiers: count(base), est_cost: bCost }, candidate: { tiers: count(cand), est_cost: cCost }, moved_down: down, moved_up: up, unchanged: entries.length - down - up, saving: bCost > 0 ? 1 - cCost / bCost : null, held_by_override: held, moves };
}

File: python/tiershift/analytics.py
"""Aggregations over the decision log for `tiershift report` and `tiershift tune`. Pure functions. Port of src/analytics.ts."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping, Optional, Sequence

from .policy import apply_policy, estimate_output_tokens
from .types import CodeSignals, Config, ModelMeta, Signals

JEV_PRICE = 0.042 / 1e6


def _pct(xs: Sequence[float], p: float) -> float:
    if not xs:
        return 0
    s = sorted(xs)
    return s[min(len(s) - 1, int(p * len(s)))]


def est_cost_for(meta: Optional[ModelMeta], in_tok: int, out_tok: int) -> Optional[float]:
    if not meta or "price" not in meta:
        return None
    price = meta["price"]
    return (in_tok * price["input"] + out_tok * price["output"]) / 1e6


def entry_cost(e: Mapping[str, Any]) -> float:
    """Cost of one entry: actual when a model was called, else the estimate."""
    if e.get("cost_usd") is not None:
        return float(e["cost_usd"])
    if e.get("est_cost_usd") is not None:
        return float(e["est_cost_usd"])
    return 0.0


def model_for_tier(config: Config, tier: str, tiers: Optional[Mapping[str, Sequence[str]]] = None) -> Optional[str]:
    lst = (tiers if tiers is not None else config["tiers"]).get(tier) or []
    return lst[0] if lst else None


@dataclass
class TierStat:
    tier: str
    n: int
    share: float
    cost: float
    est_cost: float
    mean_confidence: float
    p50_jev_ms: float
    estimate_check: Optional[dict[str, float]] = None
    """Present when the tier has entries with actual usage: n, est, actual, ratio (est / actual)."""


@dataclass
class Report:
    n: int
    from_ts: Optional[str]
    to_ts: Optional[str]
    tiers: list[TierStat]
    models: list[dict[str, Any]]
    total_cost: float
    total_est_cost: float
    flagship_est_cost: float
    flagship_model: Optional[str]
    saving_vs_flagship: Optional[float]
    jev: dict[str, float]
    degraded: int
    fell_back: int
    low_confidence: int
    overrides: list[dict[str, Any]] = field(default_factory=list)


def _out_tokens(e: Mapping[str, Any]) -> int:
    return int(e["output_tokens"]) if e.get("output_tokens") is not None else estimate_output_tokens(float(e["signals"]["output_length"]))


def build_report(entries: Sequence[Mapping[str, Any]], config: Config, tiers: Optional[Mapping[str, Sequence[str]]] = None, baseline_model: Optional[str] = None) -> Report:
    order = list(config["tiers"].keys())
    flagship_model = baseline_model or model_for_tier(config, order[-1], tiers)
    n = len(entries)
    tier_stats: list[TierStat] = []
    for tier in order:
        es = [e for e in entries if e.get("tier") == tier]
        stat = TierStat(
            tier=tier, n=len(es), share=len(es) / (n or 1),
            cost=sum(entry_cost(e) for e in es), est_cost=sum(float(e.get("est_cost_usd") or 0) for e in es),
            mean_confidence=(sum(float(e["confidence"]) for e in es) / len(es)) if es else 0,
            p50_jev_ms=_pct([float(e["jev_latency_ms"]) for e in es], 0.5),
        )
        # Where a model was actually called, compare the pre-call estimate with the billed cost so users can see the estimate error.
        both = [e for e in es if e.get("cost_usd") is not None and e.get("est_cost_usd") is not None]
        if both:
            est = sum(float(e["est_cost_usd"]) for e in both)
            actual = sum(float(e["cost_usd"]) for e in both)
            stat.estimate_check = {"n": len(both), "est": est, "actual": actual, "ratio": (est / actual) if actual > 0 else 0}
        tier_stats.append(stat)
    model_ids = sorted({e["model"] for e in entries})
    models = []
    for mid in model_ids:
        es = [e for e in entries if e["model"] == mid]
        models.append({"model": mid, "n": len(es), "cost": sum(entry_cost(e) for e in es), "est_cost": sum(float(e.get("est_cost_usd") or 0) for e in es)})
    meta = (config.get("models") or {}).get(flagship_model) if flagship_model else None
    flagship_est = sum((est_cost_for(meta, int(e["code_signals"]["est_input_tokens"]), _out_tokens(e)) or 0) for e in entries) if flagship_model else 0.0
    total = sum(entry_cost(e) for e in entries)
    counts: dict[str, int] = {}
    for e in entries:
        for r in e.get("reason", []):
            if str(r).startswith("override"):
                key = str(r).split(" → ")[0]
                counts[key] = counts.get(key, 0) + 1
    return Report(
        n=n, from_ts=entries[0]["ts"] if entries else None, to_ts=entries[-1]["ts"] if entries else None,
        tiers=tier_stats, models=models, total_cost=total, total_est_cost=sum(float(e.get("est_cost_usd") or 0) for e in entries),
        flagship_est_cost=flagship_est, flagship_model=flagship_model,
        saving_vs_flagship=(1 - total / flagship_est) if flagship_est > 0 else None,
        jev={"p50_ms": _pct([float(e["jev_latency_ms"]) for e in entries], 0.5), "p95_ms": _pct([float(e["jev_latency_ms"]) for e in entries], 0.95), "cost": sum(float(e["jev_input_tokens"]) * JEV_PRICE for e in entries)},
        degraded=sum(1 for e in entries if e.get("degraded")), fell_back=sum(1 for e in entries if e.get("fell_back")),
        low_confidence=sum(1 for e in entries if float(e["confidence"]) < 0.5),
        overrides=sorted(({"reason": k, "n": v} for k, v in counts.items()), key=lambda x: -x["n"]),
    )


@dataclass
class TuneResult:
    n: int
    baseline: dict[str, Any]
    candidate: dict[str, Any]
    moved_down: int
    moved_up: int
    unchanged: int
    saving: Optional[float]
    held_by_override: int
    moves: list[dict[str, Any]]


def tune(entries: Sequence[Mapping[str, Any]], baseline: Config, candidate: Config, tiers: Optional[Mapping[str, Sequence[str]]] = None) -> TuneResult:
    """Replay logged signals against a candidate config. No Jev calls: the signals are already in the log."""
    order = list(candidate["tiers"].keys())

    def count(ts: Sequence[str]) -> dict[str, int]:
        return {t: sum(1 for x in ts if x == t) for t in order}

    def cost(cfg: Config, e: Mapping[str, Any], tier: str) -> float:
        model = model_for_tier(cfg, tier, tiers)
        meta = (cfg.get("models") or {}).get(model) if model else None
        return est_cost_for(meta, int(e["code_signals"]["est_input_tokens"]), _out_tokens(e)) or 0.0

    def rules_only(cfg: Config) -> Config:
        return {**cfg, "overrides": []}  # type: ignore[return-value]

    base: list[str] = []
    cand: list[str] = []
    moves: list[dict[str, Any]] = []
    b_cost = c_cost = 0.0
    down = up = held = 0
    for e in entries:
        sig = Signals.from_dict(e["signals"])
        code = CodeSignals.from_dict(e["code_signals"])
        b = apply_policy(baseline, sig, code).tier
        c = apply_policy(candidate, sig, code).tier
        base.append(b)
        cand.append(c)
        b_cost += cost(baseline, e, b)
        c_cost += cost(candidate, e, c)
        if b != c:
            if order.index(c) < order.index(b):
                down += 1
            else:
                up += 1
            moves.append({"ts": e["ts"], "from": b, "to": c, "difficulty": sig.difficulty, "stakes": sig.stakes, "confidence": float(e["confidence"])})
        elif apply_policy(rules_only(baseline), sig, code).tier != apply_policy(rules_only(candidate), sig, code).tier:
            held += 1
    return TuneResult(
        n=len(entries), baseline={"tiers": count(base), "est_cost": b_cost}, candidate={"tiers": count(cand), "est_cost": c_cost},
        moved_down=down, moved_up=up, unchanged=len(entries) - down - up,
        saving=(1 - c_cost / b_cost) if b_cost > 0 else None, held_by_override=held, moves=moves,
    )
